### PR TITLE
fix(data-table): require additional props in certain situations

### LINKED ISSUE #5681: Addition of prop types to TableExpandHeader lead to invalid prop warnings
Over in: https://github.com/carbon-design-system/carbon/commit/f3422c43819222edf20a2b431414bcc04a8d19e5 we introduced prop types for `TableExpandHeader`, however, this component is typically used without any props in our examples and usage guidance. As a result, individuals who have updated to the latest are seeing prop warnings for now required props that aren't necessarily required.

We should double-check our prop types for this component and see if we need to add conditional validation for these required props for the default use-case of `<TableExpandHeader />`

### PR DESCRIPTION
**Pull request by tw15egan**
Closes https://github.com/carbon-design-system/carbon/issues/5681

Adds in a `PropType` checker function to only require certain props in certain situations

#### Changelog

**New**

- Helper function to validate if a prop is required or not. 

#### Testing / Reviewing

Ensure Data Table w/ expansion does not throw any console warnings.

### CODE REVIEW

File: packages/react/src/components/DataTable/TableExpandHeader.js [hunk 1]
```diff
@@ -49,6 +49,21 @@ const TableExpandHeader = ({
   );
 };
 
+function isRequiredIfPropExists(requiredProp) {
+  function checker(props, propName, componentName) {
+    if (props[requiredProp] === undefined) {
```
Comment (asudoh) [inline packages/react/src/components/DataTable/TableExpandHeader.js]: Thank you for jumping in @tw15egan - Tried to add `console.log('requiredProp:', requiredProp); console.log('propName:', propName)`, where `requiredProp` and `propName` seem to be always the same. So wondering what `isRequiredIfPropExists()` is meant for... Any more details? Thanks!

File: packages/react/src/components/DataTable/TableExpandHeader.js [hunk 2]
```diff
@@ -57,18 +72,18 @@ TableExpandHeader.propTypes = {
...
    */
-  ariaLabel: PropTypes.string.isRequired,
+  ariaLabel: isRequiredIfPropExists('ariaLabel'),
 
   /**
    * Specify whether this row is expanded or not. This helps coordinate data
    * attributes so that `TableExpandRow` and `TableExapndedRow` work together
    */
-  isExpanded: PropTypes.bool.isRequired,
+  isExpanded: isRequiredIfPropExists('isExpanded'),
```
Comment (joshblack) [inline packages/react/src/components/DataTable/TableExpandHeader.js]: Would this be like:

```suggestion
  isExpanded: isRequiredIfPropExists('enableExpando'),
```

?

Comment (tw15egan) [reply packages/react/src/components/DataTable/TableExpandHeader.js]: Oh right, let me fix

File: packages/react/src/components/DataTable/TableExpandHeader.js (hunk 1 above)
Comment (tw15egan) [reply packages/react/src/components/DataTable/TableExpandHeader.js]: Had a typo, should be checking if `enableExpando` exists, then those props are required

File: packages/react/src/components/DataTable/TableExpandHeader.js [hunk 3]
```diff
@@ -57,18 +72,18 @@ TableExpandHeader.propTypes = {
    * Specify the string read by a voice reader when the expand trigger is
    * focused
    */
-  ariaLabel: PropTypes.string.isRequired,
+  ariaLabel: isRequiredIfPropExists('ariaLabel'),
```
Comment (asudoh) [inline packages/react/src/components/DataTable/TableExpandHeader.js]:
```suggestion
  ariaLabel: isRequiredIfPropExists('enableExpando'),
```

File: packages/react/src/components/DataTable/TableExpandHeader.js [hunk 4]
```diff
@@ -49,6 +49,21 @@ const TableExpandHeader = ({
   );
 };
 
+function isRequiredIfPropExists(requiredProp) {
```
Comment (asudoh) [inline packages/react/src/components/DataTable/TableExpandHeader.js]: Just wanted to see if you want to re-use: https://github.com/carbon-design-system/carbon/blob/v10.10.0/packages/react/src/prop-types/requiredIfGivenPropExists.js

Comment (tw15egan) [reply packages/react/src/components/DataTable/TableExpandHeader.js]: 🤦‍♂ yes, I'll definitely refactor this to use the existing function!

**Comment by tw15egan** [comment]
@asudoh @joshblack updated and ready for re-review 👍

**Review by joshblack — APPROVED** [review]

**Review by asudoh — APPROVED** [review]
Thanks @tw15egan for the update!

**Review by aledavila — APPROVED** [review]

**Pull request merged**